### DISCUSSION TITLE
Problems with Reflection Grating Simulation

### DISCUSSION THREAD

**Discussion by Harry8251** [NanoComp/meep#2772] [Q&A] [answered]
Hello Meep Community,

I'm using Meep version 1.28.0 on Ubuntu 20 LTS via the Windows Subsystem for Linux to simulate the reflection spectrum of a grating.

Based on the great example "Diffraction Spectrum of a Binary Grating", where the transmission spectrum of a grating is calculated, I am trying to simulate the diffractive reflection spectrum of a single frequency source.

Here's a brief overview of my approach:

1.    I placed an Eigenmode Monitor within the glass grating.
2.    Conducted a normalization run with the entire cell composed of glass.
3.    Used sim.load_minus_flux_data function for my mode monitor in the run with the actual grating geometry
4.    Executed the simulation with the actual grating geometry.
5.    Evaluated the reflection coefficients per diffraction order following the methodology in the "Diffraction Spectrum of a Binary Grating" example.

![grafik](https://github.com/NanoComp/meep/assets/157973565/6c867189-9c1f-4b8c-bf70-744ed8566e85)
**Figure1** The grating height corresponds to 1/4 lambda (in medium!), so that in the reflection the zeroth and all even reflection orders should be canceled.

However, I've encountered two issues:

Problem 1:
With the grating height (gh) set to zero, I anticipated replicating the Fresnel reflection coefficient of 0.040  (n=1.5 and theta=0.0) for the zeroth diffraction order. Instead, my simulation yields a coefficient of only 0.0387.

Problem 2:
Setting the grating height (gh) to 0.333x0.25, which corresponds to half a wavelength shift in the glass medium for n=1.5 and wl=0.5. I expected the zeroth and all even diffraction orders to be nearly zero, with orders 1 and 3 carrying most of the intensity. While the simulation effectively reduces the zeroth order, the results for the other orders don't match my expectations.

Diffr. Order	Wavelegth	Intensity
grating0:,        0.50000, 0.00000164
grating1:,        0.50000, 0.00026158
grating2:,        0.50000, 0.01331711
grating3:,        0.50000, 0.00000123
grating4:,        0.50000, 0.00000045

I am familiar with the FAQ topic "Why are my reflectance/transmittance values less than zero and/or greater than one?", so I expect user error on my part. I look forward to any hints.

Thanks
Harry

```
import meep as mp
import math
import numpy as np
import matplotlib.pyplot as plt

mp.verbosity(1)
resolution = 120  # pixels/μm
runtime = 100

glass = mp.Medium(index= 1.5 )
air = mp.Medium(index=1.0)

nfreq = 1

dpml = 1.0  # PML thickness
dsub = 3.0  # substrate thickness
dpad = 1.0  # padding between grating and PML
gp = 10.0  # grating period
#gh = 0.0
gh = 0.333333 * 0.25  # grating height
gdc = 0.5  # grating duty cycle

sx = dpml + dsub + gh + dpad + dpml
sy = gp

cell_size = mp.Vector3(sx, sy, 0)
pml_layers = [mp.PML(thickness=dpml, direction=mp.X)]

wvl = 0.5  # center wavelength
fcen = 1 / wvl  # center frequency
df = 0.01  # frequency width

src_pt = mp.Vector3(-1.5, 0, 0)
sources = [mp.Source(mp.ContinuousSource(frequency=fcen, start_time=0, end_time=50),
                     component=mp.Ez,
                     center=src_pt,
                     size=mp.Vector3(0, sy, 0))
           ]

k_point = mp.Vector3(0,0,0)

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=pml_layers,
                    default_material=glass,
                    k_point=k_point,
                    sources=sources,
                    symmetries=[mp.Mirror(mp.Y)])

mon_pt = mp.Vector3(0.0, 0, 0)
flux_mon = sim.add_mode_monitor(
    fcen,
    0,
    1,
    mp.FluxRegion(center=mon_pt, size=mp.Vector3(0, sy, 0)),
)

sim.run(until=runtime)

input_flux = mp.get_fluxes(flux_mon)
input_flux_data = sim.get_flux_data(flux_mon)

sim.reset_meep()

geometry = [mp.Block(material=glass,
                      size=mp.Vector3(dpml + dsub, mp.inf, mp.inf),
                      center=mp.Vector3(-0.5 * sx + 0.5 * (dpml + dsub), 0, 0)),
              mp.Block(material=glass,
                      size=mp.Vector3(gh,gdc*gp,mp.inf),
                      center=mp.Vector3(-0.5*sx+dpml+dsub+0.5*gh,0,0))
             ]

sim = mp.Simulation(resolution=resolution,
                    cell_size=cell_size,
                    boundary_layers=pml_layers,
                    geometry=geometry,
                    k_point=k_point,
                    sources=sources,
                    symmetries=[mp.Mirror(mp.Y)])

mode_mon = sim.add_mode_monitor(
    fcen,
    0,
    1,
    mp.FluxRegion(center=mon_pt, size=mp.Vector3(0, sy, 0)),
)

sim.load_minus_flux_data(mode_mon, input_flux_data)

sim.run(until=runtime)

freqs = mp.get_eigenmode_freqs(mode_mon)

nmode = 5
res = sim.get_eigenmode_coefficients(mode_mon, range(1, nmode + 1), eig_parity=mp.ODD_Z)
coeffs = res.alpha

mode_wvl = []
mode_tran = []

for nm in range(nmode):
    for nf in range(nfreq):
        mode_wvl.append(1 / freqs[nf])
        tran = abs(coeffs[nm, nf, 1]) ** 2 / input_flux[nf]
        mode_tran.append(0.5 * tran if nm != 0 else tran)
        print("grating{}:, {:.5f}, {:.8f}".format(nm, mode_wvl[-1], mode_tran[-1]))


```

**Comment by oskooi** [comment, marked as answer]
A couple of suggestions:

1. For calculations involving the DFT fields (required for mode decomposition), you should be using a *pulsed* (i.e., `GaussianSource`) rather than a CW source. See this [FAQ](https://meep.readthedocs.io/en/latest/FAQ/#for-calculations-involving-fourier-transformed-fields-why-should-the-source-be-a-pulse-rather-than-a-continuous-wave).
2. In Problem 1, the simulated results are close but not quite equal to the expected result. This may indicate that your simulation is not yet converged.  There are two things you can try: (1) increase the resolution or (2) run for a longer time (using a pulsed source) to ensure the fields have fully decayed away. See this [FAQ](https://meep.readthedocs.io/en/latest/FAQ/#checking-convergence).

Also, note that when performing mode decomposition at the end of your simulation, you can specify the diffraction orders using a [`DiffractedPlanewave`](https://meep.readthedocs.io/en/latest/Python_User_Interface/#diffractedplanewave) object as demonstrated in [Tutorial/Diffracted Planewaves in Homogeneous Media](https://meep.readthedocs.io/en/latest/Python_Tutorials/Mode_Decomposition/#diffracted-planewaves-in-homogeneous-media) rather than a band number and `eig_parity`.

**Reply by Harry8251** [reply]
Hi Oskooi,

Thanks for your suggestions. I have successfully implemented them and also had a look at the recently updated test case test_binary_grating.py.

**Comment by 10UAikawa** [comment]
How do you make those graphical images for your structure? Do you use Meep to do it?
